### PR TITLE
Feature/adapt vimeo

### PR DESCRIPTION
**Pull request by danielstorey**

### CODE REVIEW

File: example.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    "_classes": "",
+    "_layout": "full",
+    "_component": "vimeo",
+    "title": "Component Title",
+    "displayTitle": "Component Title",
+    "body": "This is a vimeo component",
+    "instruction":"Click the button to play",
+    "_setCompletionOn": "play",
+    "_media": {
+        "_source": "https://vimeo.com/319977833",
```
Comment (moloko) [inline example.json]: might be better to use a protocol-less url here i.e. `"_source": "//vimeo.com/319977833"`

File: example.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
+    "_id": "c-35",
+    "_parentId":"b-20",
```
Comment (moloko) [inline example.json]: needs a space between the colon and the opening quote (also in `_type` and `instruction`)

File: templates/vimeoTranscript.hbs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+                {{else}}
+                    {{transcriptLink}}
+                {{/if}}
+            </div>
+        </button>
+    {{/if}}
+</div>
+
+{{#if _inlineTranscript}}
+    <div class="media-inline-transcript-body-container" role="region" aria-label="{{#if inlineTranscriptButton}}{{inlineTranscriptButton}}{{else}}{{transcriptLink}}{{/if}}">
```
Comment (moloko) [inline templates/vimeoTranscript.hbs]: the `else` here needs removing as the logic is currently that if there's no inlineTranscriptButton defined use the link to the external transcript as the ARIA label for the inline transcript div... can't see how that would ever something you'd actually want!

 (this is also an issue in the adapt media component which will be sorted as part of v5)

File: templates/vimeoTranscript.hbs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+                {{#if inlineTranscriptButton}}
+                    {{inlineTranscriptButton}}
+                {{else}}
+                    {{transcriptLink}}
+                {{/if}}
+            </div>
+        </button>
+    {{/if}}
+    {{#if _externalTranscript}}
+        <button onclick="window.open('{{transcriptLink}}')" class="media-external-transcript-button button">
```
Comment (moloko) [inline templates/vimeoTranscript.hbs]:
```suggestion
        <button onclick="top.window.open('{{transcriptLink}}')" class="media-external-transcript-button button" role="link">
```

File: templates/vimeoTranscript.hbs [hunk 5]
```diff
@@ -0,0 +1,32 @@
+<div class="media-transcript-button-container">
+    {{#if _inlineTranscript}}
+        <button class="media-inline-transcript-button button" aria-expanded="false">
+            <div class="transcript-text-container">
+                {{#if inlineTranscriptButton}}
+                    {{inlineTranscriptButton}}
+                {{else}}
```
Comment (moloko) [inline templates/vimeoTranscript.hbs]: drop the `else` block, can't see why you'd ever want the external transcript url to be shown as the label for the inline transcript button!

File: js/VimeoView.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+         * and set the padding on this view's element accordingly
+         */
+        setupResponsiveSizing: function() {
+            var $el = this.$el;
+            var player = this.player;
+            var self = this;
+
+            player
+                .getVideoWidth()
+                .then(function(width) {
```
Comment (moloko) [inline js/VimeoView.js]: will this work in IE?

File: properties.schema [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+  "$ref": "http://localhost/plugins/content/component/model.schema",
+  "globals": {
+    "ariaRegion": {
+      "type": "string",
+      "required": true,
+      "default": "This is a media component which displays a Vimeo video. Select the play / pause button to watch it.",
+      "inputType": "Text",
+      "validators": []
+    },
+    "transcriptButton": {
```
Comment (moloko) [inline properties.schema]: This shouldn't be here... also I think you're missing `skipToTranscript` so maybe change to to that?

File: properties.schema [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+      "default": "both",
+      "editorOnly": true
+    },
+    "instruction": {
+      "type": "string",
+      "required": false,
+      "default": "",
+      "inputType": "Text",
+      "validators": [],
+      "help": "This is the instruction text"
```
Comment (moloko) [inline properties.schema]: I think you can get rid of this help text

File: properties.schema [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    },
+    "_setCompletionOn": {
+      "type": "string",
+      "required": true,
+      "enum": ["inview", "play", "ended"],
+      "default": "inview",
```
Comment (moloko) [inline properties.schema]: Change this to 'play' to match example.json?

File: properties.schema [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+      "title": "Set completion of Video on",
+      "inputType": {"type": "Select", "options":["inview", "play", "ended"]},
+      "validators": ["required"],
+      "help": "This tells Adapt when to set this video to complete"
```
Comment (moloko) [inline properties.schema]:
```suggestion
      "help": "Defines what media event should trigger completion of this component."
```

File: templates/vimeo.hbs [hunk 11]
```diff
@@ -0,0 +1,4 @@
+<div class="component-inner vimeo-inner">
+    {{> component this}}
```
Comment (moloko) [inline templates/vimeo.hbs]:
```suggestion
    {{> component this}}
    {{#any _transcript._externalTranscript _transcript._inlineTranscript}}
        <button class="aria-label js-skip-to-transcript" aria-label="{{_globals._components._vimeo.skipToTranscript}}"></button>
    {{/any}}
```

File: js/VimeoView.js (hunk 6 above)
Comment (danielstorey) [reply js/VimeoView.js]: Fair point, and TBH I should have checked before but looks like the Vimeo library ships with its own Promise polyfill. Tested in IE11 and it works

File: README.md [hunk 12]
```diff
@@ -1 +1,66 @@
...
+
+**instruction** (string): This optional text appears above the component. It is frequently used to
+guide the learner’s interaction with the component.
+
+**_setCompletionOn** (string): This determines when Adapt will register this component as having been completed by the user. Acceptable values are `inview` (triggered when the component is fully displayed within the viewport), `play` (triggered when playback is initiated), or `ended` (triggered when the video has reached the end of playback). 
+
+
+**_media** (object): The media configuration, containing values for **_source**, **_autoplay** and **_loop**
+
+>**__source** (string): The URL of the Vimeo video. This can be a direct link or an embed link
```
Comment (moloko) [inline README.md]: additional underscore here

**Comment by oliverfoster** [comment]
Is it necessary to have the vimeo player in the libraries folder, as opposed to it being downloaded from a cdn? 

Any advantages?

https://developer.vimeo.com/player/sdk/basics
https://github.com/vimeo/player.js

**Comment by danielstorey** [comment]
> Is it necessary to have the vimeo player in the libraries folder, as opposed to it being downloaded from a cdn?
> 
> Any advantages?
> 
> https://developer.vimeo.com/player/sdk/basics
> https://github.com/vimeo/player.js

Fair point. Maybe including the library provides some protection against possible future updates which might break stuff?!

**Comment by moloko** [comment]
@oliverfoster we have a few clients who require that e-learning content is completely self-contained i.e. doesn't use any code or assets loaded from external servers. So it would be preferable for it to be in the libraries folder from that point of view.

**Comment by oliverfoster** [comment]
@moloko I'm not sure if that's possible in this situation as the video must be loaded from vimeo, which is an external platform.

@danielstorey I guess it depends which changes would cause the break? If the client code changes to suit a new server-side API then having the client code fixed against a redundant server-side API would be a problem. Equally, if the client side code changes significantly enough to break Adapt, that isn't good either.

**Comment by moloko** [comment]
> I'm not sure if that's possible in this situation as the video must be loaded from vimeo, which is an external platform.

oh yeah! derp ;-)

**Comment by oliverfoster** [comment]
I happy to leave as is. I just wanted to ask. 👍

**Pull request merged**